### PR TITLE
VMA Fixes

### PR DESCRIPTION
**Pull request by JSCU-CNI**
Two bugs related to VMA are covered in this PR:

- vma-extract had a seek to the start of the data inside an extent in the wrong place (within the loop over the blocks instead of before the loop) 
- A logic-bug related to the fact  that inside an extent, blocks of multiple devices could be stored. These blocks should be handled in the loop inside `_iter_clusters` as well, in order to return the right `block_offset`.

### CODE REVIEW

File: dissect/hypervisor/backup/vma.py [hunk 1]
```diff
@@ -164,7 +163,7 @@ def __init__(self, fh, offset):
             elif cluster_num > self._max[dev_id]:
                 self._max[dev_id] = cluster_num
 
-            self.blocks[dev_id].append((dev_id, cluster_num, mask))
+            self.blocks.append((dev_id, cluster_num, mask))
```
Comment (Schamper) [inline dissect/hypervisor/backup/vma.py:166]: I don't yet see how this change affects anything, can you walk me through it or provide an example with e.g. some mock tuples?

In the snippet later on, you iterate over all entries in `extent.blocks`, does that not achieve the same as what the original code did, with the "filtering" being done in the dictionary? You write "inside an extent, blocks of multiple devices could be stored", but doesn't it already support that by storing those blocks in separate dictionary keys?

**Review by Schamper — COMMENTED** [review]
Also small suggestion (can't put a comment there), can you change the `@lru_cache(65536)` at line 94 to a `self._extent = lru_cache(65536)(self._extent)` at the bottom of `__init__`?

**Comment by JSCU-CNI** [comment]
> Also small suggestion (can't put a comment there), can you change the `@lru_cache(65536)` at line 94 to a `self._extent = lru_cache(65536)(self._extent)` at the bottom of `__init__`?

Done! Could you please explain your reasons for asking this? We don't see how this benefits the code, but we suspect you've got good reasons for doing so :smile:

File: dissect/hypervisor/backup/vma.py (hunk 1 above)
Comment (JSCU-CNI) [reply dissect/hypervisor/backup/vma.py:167]: Sorry for any confusion that a lack of explanation from our side may have caused. Image the following extent: 
```
|--------|------------|--------|
| dev_id | cluster_no | mask   |
|--------|------------|--------|
|      1 |          0 | 0xFFFF |
|      1 |          1 | 0xFFFF |
|      2 |          0 | 0xFFFF |
|      1 |          2 | 0xFFFF |
|    ... |        ... |    ... |
|--------|------------|--------|
```
The function `_iter_clusters(vma, device=2, cluster=0, count=1)` should yield `(0, 0xFFFF, 2 * 16 * c_vma.VMA_BLOCK_SIZE)`. The old code returned `(0, 0xFFFF, 0)` instead, since the `block_offset` was not incremented for the first two clusters in this extent (because the loop only iterated over `extent.blocks[dev_id]`, thus missing the first blocks).

File: dissect/hypervisor/backup/vma.py [hunk 2]
```diff
@@ -67,6 +67,8 @@ def __init__(self, fh):
 
             self._devices[dev_id] = Device(self, dev_id, self.blob_string(dev_info.devname_ptr), dev_info.size)
 
+        self.extent = @lru_cache(65536)(self.extent)
```
Comment (Schamper) [inline dissect/hypervisor/backup/vma.py:70]:
```suggestion
        self._extent = @lru_cache(65536)(self._extent)
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by Schamper** [comment]
> > Also small suggestion (can't put a comment there), can you change the `@lru_cache(65536)` at line 94 to a `self._extent = lru_cache(65536)(self._extent)` at the bottom of `__init__`?
> 
> Done! Could you please explain your reasons for asking this? We don't see how this benefits the code, but we suspect you've got good reasons for doing so smile

https://rednafi.github.io/python/lru_cache_on_methods/

TL;DR: `@lru_cache` on class methods is bad, as it can result in an unintended memory leak because `self` is cached.

File: dissect/hypervisor/backup/vma.py (hunk 2 above)
Comment (JSCU-CNI) [reply dissect/hypervisor/backup/vma.py]: :see_no_evil:

File: dissect/hypervisor/backup/vma.py (hunk 1 above)
Comment (Schamper) [reply dissect/hypervisor/backup/vma.py:167]: Aaaah, the devil is in the `block_offset` calculation, I see it now, thanks!

I then wonder if this code would be simpler if we moved the `block_offset` calculation to the loop in the `Extent` class. What's your opinion on that?

Comment (JSCU-CNI) [reply dissect/hypervisor/backup/vma.py:167]: Sounds reasonable! This would also make the results of the offset calculations being stored in the LRU cache, isn't it?

Comment (Schamper) [reply dissect/hypervisor/backup/vma.py:167]: Yes, I believe so!

File: dissect/hypervisor/backup/vma.py [hunk 3]
```diff
@@ -206,20 +215,19 @@ def _read(self, offset, length):
         return b"".join(result)
 
 
-def _iter_clusters(vma, dev_id, cluster, count):
+def _iter_clusters(vma, device, cluster, count):
```
Comment (Schamper) [inline dissect/hypervisor/backup/vma.py:218]: The change from `dev_id` to `device` is not really necessary, it makes it a little less clear IMO.

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]
